In [1]:
import pandas as pd
import numpy as np

In [7]:
data=pd.read_csv(r"C:\Users\yash.jha\Downloads\Mercedes.csv",usecols=['X1','X2','X3','X4','X5','X6'])

In [8]:
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [9]:
#Let's check how many unique categories are present in each column
for col in data.columns:
    print(col, ": ",len(data[col].unique()), 'labels')

X1 :  27 labels
X2 :  44 labels
X3 :  7 labels
X4 :  4 labels
X5 :  29 labels
X6 :  12 labels


In [10]:
#let's examine how many columns we will obtain after one hot encoding these variables
pd.get_dummies(data, drop_first=True).shape

(4209, 117)

We can see that from just 6 initial categorical variables, We ended up with 117 new variables

What can we do instead?
In the winning solution of the KDD 2009 cup: "Winning the KDD cup Orange challange with Ensemble selection the authors limit one hot encoding to the 10 most frequent labels of the variable. This means that they would make one binary variable for each of the 10 most frequent labels only. This is equivalent to grouping all the other labels under a new category, that in this case will be dropped. Thus, the 10 new dummy variables indicate if one of the 10 most frequent labels is present in (1) or not(0) for a particular observation."

How can we do that in python/

In [11]:
#Let's find the top 10 most frequent categories for the variable x2
data.X2.value_counts().sort_values(ascending=False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
ag      19
z       19
Name: X2, dtype: int64

In [12]:
#We are going to take the top 10 values
# Let's make a list with the most frequent categories of the variable
top_10=[X for X in data.X2.value_counts().sort_values(ascending=False).head(10).index]
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [14]:
# and now we make the 10 binary variables
for label in top_10:
    data[label]=np.where(data['X2']==label,1,0)
    
data[['X2']+top_10].head(40)

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0
20   m   0   0   0  1   0  0  0  0  0  0
21  as   1   0   0  0   0  0  0  0  0  0
22   a   0   0   0  0   0  0  0  0  0  0
23   m   0   0   0  1   0  0  0  0  0  0
24  ak   0   0   0  0   1  0  0  0  0  0
25   m   0   0   0  1   0  0  0  0  0  0
26  as   1   0   0  0   0  0  0  0  0  0
27  as   1   0   0  0   0  0  0  0  0  0
28  as   1   0   0  0   0  0  0  0  0  0
29   m   0   0   0  1   0  0  0  0  0  0
30  as   1   0   0  0   0  0  0  0  0  0
31  as   1   0   0  0   0  0  0  0  0  0
32   m   0   0   0  1   0  0  0  0  0  0
33  as   1   0   0  0   0  0  0  0  0  0
34  as   1   0   0  0   0  0  0  0  0  0
35  ai   0   0   1  0   0  0  0  0  0  0
36  as   1   0   0  0   0  0  0  0  0  0
37   k   0   0   0  0   0  0  0  0  0  0
38  as   1   0   0  0   0  0  0  0  0  0
39  as   1   0   0  0   0  0  0  0  0  0

In [16]:
# get whole set of dummy variables, for all the categorical variables

def one_hot_top_x(df, variable, top_x_labels):
    #function to create dummy variables for the most frequent labels
    #We can Vary the number of most frequent labels that we encode
    
    for label in top_x_labels:
        df[variable+'_'+label]=np.where(data[variable]==label, 1, 0)
        
#read the data again
data=pd.read_csv(r"C:\Users\yash.jha\Downloads\Mercedes.csv",usecols=['X1','X2','X3','X4','X5','X6'])

#encode X2 into the 10 most frequent categories
one_hot_top_x(data,'X2',top_10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0  v  at  a  d  u  j      0      0      0     0      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0      0     0     1     0   
3  t   n  f  d  x  l      0      0      0     0      0     0     1     0   
4  v   n  f  d  h  d      0      0      0     0      0     0     1     0   

   X2_f  X2_e  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0

In [18]:
#find the 10 most frequent categories for X1
top_10=[X for X in data.X1.value_counts().sort_values(ascending=False).head(10).index]

#Now create the 10 most frequent dummy variables for X1
one_hot_top_x(data,'X1',top_10)
data.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X1_aa  X1_s  X1_b  X1_l  \
0  v  at  a  d  u  j      0      0      0     0  ...      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0  ...      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0  ...      0     0     0     0   
3  t   n  f  d  x  l      0      0      0     0  ...      0     0     0     0   
4  v   n  f  d  h  d      0      0      0     0  ...      0     0     0     0   

   X1_v  X1_r  X1_i  X1_a  X1_c  X1_o  
0     1     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     0     0     0     0  
3     0     0     0     0     0     0  
4     1     0     0     0     0     0  

[5 rows x 26 columns]

## One Hot Encoding of top Variables

### Advantages
- Straightforward to implement
- Does not required hrs of variable exploration
- Does not expand massively the feature space (number the columns in the dataset)

### Disadvantages
- Does not add any information that may make the variable more predictive
- Does not keep the information of the ignored labels



Because it is not unusual that categorical variables have a few dominating categories and the remaining labels add mostly noise, 
This is a quite simple and straightforward approach that may be useful on many occasions.

